src/source.rs

```diff
@@ -0,0 +1,347 @@
+use std::{
+  borrow::Cow,
+  convert::{TryFrom, TryInto},
+  fmt,
+  hash::{Hash, Hasher},
+};
+
+use serde::{Deserialize, Serialize};
+
+use crate::{
+  helpers::{decode_mappings, StreamChunks},
+  Result,
+};
+
+pub type BoxSource = Box<dyn Source>;
+
+pub trait Source: StreamChunks + DynHash + fmt::Debug + Sync + Send {
+  fn source(&self) -> Cow<str>;
+
+  fn buffer(&self) -> Cow<[u8]>;
+
+  fn size(&self) -> usize;
+
+  fn map(&self, options: &MapOptions) -> Option<SourceMap>;
+
+  fn update_hash(&self, state: &mut dyn Hasher) {
+    self.dyn_hash(state);
+  }
+}
+
+impl<T: Source + 'static> From<T> for BoxSource {
+  fn from(s: T) -> Self {
+    Box::new(s)
+  }
+}
+
+// for `updateHash`
+pub trait DynHash {
+  fn dyn_hash(&self, state: &mut dyn Hasher);
+}
+
+impl<H: Hash> DynHash for H {
+  fn dyn_hash(&self, mut state: &mut dyn Hasher) {
+    self.hash(&mut state);
+  }
+}
+
+impl Hash for dyn Source {
+  fn hash<H: Hasher>(&self, state: &mut H) {
+    self.dyn_hash(state);
+  }
+}
+
+pub trait SourceExt {
+  fn boxed(self) -> BoxSource;
+}
+
+impl<T: Source + 'static> SourceExt for T {
+  fn boxed(self) -> BoxSource {
+    self.into()
+  }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct MapOptions {
+  pub columns: bool,
+  pub final_source: bool,
+}
+
+impl Default for MapOptions {
+  fn default() -> Self {
+    Self {
+      columns: true,
+      final_source: false,
+    }
+  }
+}
+
+impl MapOptions {
+  pub fn new(columns: bool) -> Self {
+    Self {
+      columns,
+      ..Default::default()
+    }
+  }
+}
+
+#[derive(Debug, Clone, Default, PartialEq, Eq, Hash)]
+pub struct SourceMap {
+  file: Option<String>,
+  mappings: String,
+  sources: Vec<String>,
+  sources_content: Vec<String>,
+  names: Vec<String>,
+}
+
+impl SourceMap {
+  pub fn new<S: Into<String>>(
```

Comment:
Static dispatch for every args seems weird to me, maybe we could split it into multiple generics?

Reply:
I will just delete this static dispatch since it is really weird and we will need `None::<String>` if use multiple generics.